### PR TITLE
feat: flag to suppress existing diagnostics

### PR DESCRIPTION
**Pull request by anthonyshew**
## Summary

Working on https://github.com/biomejs/biome/issues/4007 to create a flag for writing ignores. This is useful when migrating a large codebase, and you want to enable a rule that has hundreds (thousands?!) of existing violations for that rule. With this feature, you can have Biome ignore those existing violations, and burn them down later.

## Test Plan

Some new tests have been written in `crates/biome_cli/tests/cases/suppressions.rs`. Additionally, I've ran this against a large codebase to validate that things work as I'd expect.

### CODE REVIEW

**Comment by anthonyshew** [comment]
Would appreciate a look from a maintainer here to see if I'm on the right track architecturally. I'm not completely sure that I have the right idea. I was following what looks like the strategy for `--fix`, which seems like it follows the same workflow (find a diagnostic, make an edit).

**Review by dyc3 — COMMENTED** [review]
Definitely take a look at ematipico's comment here: https://github.com/biomejs/biome/issues/4007#issuecomment-2363404939

**Comment by anthonyshew** [comment]
TL;DR from Ema's comment: I wasn't even close on my first attempt. 😄 Circling back.

File: crates/biome_cli/src/commands/mod.rs
Comment (pacak) [reply crates/biome_cli/src/commands/mod.rs:226]: If you want to be able to pass both `--write-suppressions` to get `Some("default_val")` and `--write-suppressions=something to get `Some("something")` then yea, what you found is the right value.

Make a function `fn write_suppressions() -> impl Parser<Option<String>>` with code from the last comment (for `bar`), and do something like

```rust
        #[bpaf(external)]
        write_suppressions: Option<String>,
```

> I'd prefer if someone more familiar with bpaf weighed in.

:eyes:

File: crates/biome_service/src/file_handlers/javascript.rs
Comment (dyc3) [inline crates/biome_service/src/file_handlers/javascript.rs:689]: So this is not the actual point where the code gets modified. This is still a part of the `analyze` call above.

I think at this point you actually want to `return ControlFlow::Break(<suppression_action>)`

Fixes actually get applied below, at this point:

https://github.com/biomejs/biome/blob/a3483e4298213290eedd1df4225e5e15cfcf3972/crates/biome_service/src/file_handlers/javascript.rs#L695-L697

The way that analyzer rules apply fixes is by specifying a "mutation" for how the code gets modified. You'll need to create a mutation that suppresses the fix, and then apply that mutation here (probably?) instead of the actual fix.

You'll want to look at implementations of the `SuppressionAction` trait. Of particular interest, this is likely what you want to create that suppression action: 

https://github.com/biomejs/biome/blob/a3483e4298213290eedd1df4225e5e15cfcf3972/crates/biome_js_analyze/src/suppression_action.rs#L127-L132

(also note the `find_token_to_apply_suppression` function, you'll need that too)

You can probably extract the syntax token to apply the suppression to by inspecting `action.mutation.root()`

Comment (anthonyshew) [reply crates/biome_service/src/file_handlers/javascript.rs]: I definitely had the wrong idea for a while. 😄 I had started picking up on what you've described above but what you've written helped solidify it all. Thank you a ton!

**Comment by anthonyshew** [comment]
Well, it at least works now. 😄 Stuff I still need to solve for:
- The default message is still `<explanation>`
  - Likely will do this in a follow up PR but would love to make the explanation customizable
- The adding of the comment is dropping formatting. Things were working correctly at one point and then I introduced this later on. Need to do some bisecting.
- Take it for a larger test in a bigger repo than my test repo
- Write tests

File: crates/biome_service/src/file_handlers/javascript.rs
Comment (anthonyshew) [inline crates/biome_service/src/file_handlers/javascript.rs:668]: Hm, this isn't great.

File: crates/biome_cli/src/commands/mod.rs
Comment (anthonyshew) [reply crates/biome_cli/src/commands/mod.rs]: Ended up going with a different API. Thanks for the tips, though! 🧠

File: crates/biome_service/src/file_handlers/javascript.rs
Comment (anthonyshew) [inline crates/biome_service/src/file_handlers/javascript.rs:669]: Safe to say this isn't right, but it stopped a bunch of ignores getting written for `lint/nursery/useStrictMode`. I don't know if it's a bug with that rule because there aren't any other nursery rules getting written from what I can tell.

Comment (anthonyshew) [inline crates/biome_service/src/file_handlers/javascript.rs:669]: Safe to say this isn't right, but it stopped a bunch of ignores getting written for `lint/nursery/useStrictMode`. I don't know if it's a bug with that rule because there aren't any other nursery rules getting written from what I can tell.

File: crates/biome_service/src/file_handlers/graphql.rs
Comment (anthonyshew) [inline crates/biome_service/src/file_handlers/graphql.rs:534]: TODO: It looks to me like each language has the same code in these blocks, so I'd imagine I can bring the code from `javascript.rs`? (assuming that code is right)

File: crates/biome_cli/src/execute/mod.rs
Comment (anthonyshew) [inline crates/biome_cli/src/execute/mod.rs:310]: TODO: Implement these for the other traversal modes.

File: crates/biome_lsp/src/handlers/analysis.rs
Comment (anthonyshew) [inline crates/biome_lsp/src/handlers/analysis.rs:245]: TODO: Weird...

**Comment by dyc3** [comment]
For tests, take a look at the cli tests in `biome_cli`.

**Review by anthonyshew — COMMENTED** [review]
Looking for an early review from a maintainer here. This works in my large codebase from what I can tell, with a few caveats:
- `biome lint .` takes 2 seconds while `biome lint . --suppress` takes 30 seconds. I didn't expect it to be that slow...Maybe disk IO? Even that would be surprising. Or perhaps because `javascript.rs#662` no longer does `if action.is_suppression() { continue; }`?
- I'm having trouble figuring out how to write a custom explanation into the suppression action. An IDE recommendation writes with `<explanation>` and then you edit it after, so maybe this doesn't exist yet and I'm going to have to invent it?
  - This is the missing piece for me personally to be satisfied with using this on my production codebase, for what it's worth.


I've left other notes in this review, mostly as reminders to myself, but feel free to comment on those as well.

File: crates/biome_cli/src/commands/mod.rs
Comment (pacak) [reply crates/biome_cli/src/commands/mod.rs]:
```rust
    } else if !suppress && suppress_reason.is_some() {
        return Err(CliDiagnostic::IncompatibleEndConfiguration(
            IncompatibleEndConfiguration {
                reason: "--suppress-reason must be used with --suppress".to_string(),
            },
        ));
    }
```

This is one of the patterns `bpaf` is designed to prevent. You get your data from the parser and data are ready to be used without having to do additional validation and/or unwraps. As far as I understand it out of 4 possible combinations of those flags 3 are valid:

- `--suppress` alone
- `--suppress` + `--suppress-reason`
- neither is present

Natural representation for this in Rust is `Option<Option<String>>`, or inner option can be replaced with a custom enum - this reason + all reasons(?).

So

```rust
#[derive(Bpaf)]
enum Reason {
    WithReason {
        /// help here
        #[bpaf(long("suppress"))]
        _flag: (),
        suppress_reason: Option<String>,
    }
    /// help here (same as before)
    #[bpaf(long("suppress"))]
    NoReason
}
```

And use that as

```rust
    ...
    #[bpaf(external(reason))]
    write_suppressions: Option<Reason>
```

You can make enum a bit more natural with combinatoric API, but in practice this should be good enough.

File: crates/biome_service/src/file_handlers/javascript.rs
Comment (dyc3) [reply crates/biome_service/src/file_handlers/javascript.rs:669]: Yeah this is definitely weird to me. I'm assuming `useStrictMode` is disabled in your config, right? I would've thought that disabled rules wouldn't get evaluated (and therefore not emit diagnostics at all).

Comment (anthonyshew) [reply crates/biome_service/src/file_handlers/javascript.rs:669]: Correct, no nursery rules in our configuration.

**Review by dyc3 — COMMENTED** [review]
- For tests, take a look at the cli tests in `biome_cli`.
- Adding support for other languages is probably easy enough, but I'm leaning toward a new PR for that.
- It doesn't look like we have a mechanism to customize the suppression message yet: https://github.com/biomejs/biome/blob/a3483e4298213290eedd1df4225e5e15cfcf3972/crates/biome_js_analyze/src/suppression_action.rs#L219
- I ran biome with `--suppress` on a bunch of js files in bulk, and my memory usage shot up about 5 gigs in about half a second, and proceeded to consume the rest of the memory I had available within about 10 seconds (~30 Gb). I have a feeling that the perf issues are related to memory allocations for applied fixes. This might require some internal plumbing work to fix (which IMO is out of scope for this PR).

**Comment by ematipico** [comment]
>  `biome lint .` takes 2 seconds while `biome lint . --suppress` takes 30 seconds. I didn't expect it to be that slow...Maybe disk IO? Even that would be surprising. Or perhaps because `javascript.rs#662` no longer does `if action.is_suppression() { continue; }`?

We will have to look into it. However, it's safe to say that suppression actions are more involved than the other fixes, because the suppression action needs to work in a generic way. 

> I'm having trouble figuring out how to write a custom explanation into the suppression action. An IDE recommendation writes with `<explanation>` and then you edit it after, so maybe this doesn't exist yet and I'm going to have to invent it?
> This is the missing piece for me personally to be satisfied with using this on my production codebase, for what it's worth.
> 

It doesn't exist yet; that's why I suggested starting without explanation: to reduce the amount to changes, and to progressively land the feature. 

The suppression actions are implemented as a trait:

https://github.com/biomejs/biome/blob/7e29f06930e977d5f2165260ca87c21f414cf62e/crates/biome_js_analyze/src/suppression_action.rs#L50-L262

As you can notice, the reason is hardcoded, so we will have to change each struct that implements `SuppressionAction` to accept an optional reason.

File: crates/biome_service/src/file_handlers/javascript.rs
Comment (ematipico) [reply crates/biome_service/src/file_handlers/javascript.rs]: Nursery rules are enabled in debug builds, but they are turned off in production builds. So if a nursery rule is *recommended* and it has a fix, it's normal to see the changes.

**Comment by Conaclos** [comment]
> `biome lint .` takes 2 seconds while `biome lint . --suppress` takes 30 seconds. I didn't expect it to be that slow...Maybe disk IO? Even that would be surprising.

Have you tested it in release mode (`cargo build --bin biome --release` and then `./target/release/biome`)? I already got ricked by that.

File: crates/biome_service/src/file_handlers/graphql.rs
Comment (ematipico) [reply crates/biome_service/src/file_handlers/graphql.rs:534]: It's possible to make the code DRY, but be aware of the formatting step:

https://github.com/biomejs/biome/blob/7e29f06930e977d5f2165260ca87c21f414cf62e/crates/biome_service/src/file_handlers/javascript.rs#L720-L728

The `format_node` function is language specific. This is a nice to have, so we could DRY it later.

File: crates/biome_cli/src/commands/mod.rs
Comment (ematipico) [inline crates/biome_cli/src/commands/mod.rs:1]: Let's implement the new options **only** for the `lint` command:
- the `format` suppression system works differently, and can't be automated;
- `migrate` doesn't do file system traversal for now;
- `check` is a wider command, and it could in conflict with other situation e,g, `biome check --linter-enabled=false --write --suppression`
- `ci` is a read-only command, there aren't writing operations

File: crates/biome_lsp/src/handlers/analysis.rs
Comment (ematipico) [reply crates/biome_lsp/src/handlers/analysis.rs:245]: This isn't weird, but it's a normal side-effect of our architecture. Biome, internally, relies on `Workspace` trait that is used both by the CLI and LSP, which are considered as "clients" of our architecture.

File: crates/biome_service/src/file_handlers/javascript.rs
Comment (anthonyshew) [reply crates/biome_service/src/file_handlers/javascript.rs]: Oh, TIL! That explains that then. Removing this.

File: crates/biome_cli/src/execute/mod.rs
Comment (anthonyshew) [reply crates/biome_cli/src/execute/mod.rs:310]: Nope! We're doing lint only.

File: crates/biome_cli/src/commands/mod.rs
Comment (anthonyshew) [reply crates/biome_cli/src/commands/mod.rs:1]: Makes sense. Pulled back everything but `lint`.

File: crates/biome_cli/src/commands/mod.rs
Comment (anthonyshew) [inline crates/biome_cli/src/commands/mod.rs:186]:
```suggestion
        /// Suppress existing diagnostics with a `// biome-ignore` comment (Only implemented for JavaScript)
```

**Comment by anthonyshew** [comment]
Thanks for the help, all.

Status update:
- I've pulled back a bunch of changes so `--suppress` should only handle lints for JavaScript.
- I've just found out that these changes aren't causing a performance regression like I once thought. Things are still as fast as expected in my large codebase - but it appears the process isn't exiting. I have the `// biome-ignore` comments written to disk, but the process is still open, not exiting. I'll need to look into this and if anyone can verify that that see the same behavior, that'd be helpful.
- Still need to add/fix tests and the codegen check.

File: crates/biome_service/src/file_handlers/javascript.rs
Comment (anthonyshew) [inline crates/biome_service/src/file_handlers/javascript.rs:667]: This is problematic, though I can't sort out why this is. @ematipico let me know that [nursery rules are enabled in debug builds](https://github.com/biomejs/biome/pull/4008#discussion_r1782460430) but I'm seeing nursery rules get ignores written with a release build. 🤔

File: crates/biome_service/src/file_handlers/javascript.rs
Comment (anthonyshew) [inline crates/biome_service/src/file_handlers/javascript.rs:670]: This is problematic, though I'm having trouble sorting out why its happening. @ematipico let me know that [nursery rules are enabled in debug builds](https://github.com/biomejs/biome/pull/4008#discussion_r1782460430) but I'm seeing nursery rules get ignores written with a release build. 🤔

Comment (ematipico) [reply crates/biome_service/src/file_handlers/javascript.rs:669]: Oh yeah, this is something that *wasn't* written anywhere, so I took the opportunity to do that: https://github.com/biomejs/biome/blob/main/CONTRIBUTING.md#production-binaries

You'll have to provide `BIOME_VERSION` env variable to turn them off

**Comment by anthonyshew** [comment]
Alright, I think I finally have this ready! Learned a lot about how Biome works and will handle follow-ups for customizing the explanation once this one lands.

Thanks for tolerating my noob questions, @dyc3, @ematipico, and @Conaclos. 😄

File: crates/biome_cli/src/commands/mod.rs
Comment (ematipico) [inline crates/biome_cli/src/commands/mod.rs:186]:
```suggestion
        /// Fix diagnostics with suppression comments if the language supports it. 
```

I figured we could use a more broad phrasing, without citing JS

File: crates/biome_service/src/workspace.rs
Comment (ematipico) [inline crates/biome_service/src/workspace.rs:668]: Let's document this new variant

File: crates/biome_service/src/workspace.rs
Comment (ematipico) [inline crates/biome_service/src/workspace.rs:676]: I think `suppress` is not needed anymore. If `FixFileMode` already has a new variant, what's the new parameter for?

File: crates/biome_service/src/file_handlers/json.rs
Comment (ematipico) [inline crates/biome_service/src/file_handlers/json.rs:588]:
```suggestion
                        // TODO: implement once a JSON suppression action is available
```

I figured we should add a `TODO` because `JSONC` has comments, so once the analyzer supports a suppression action, we can implement it

File: crates/biome_service/src/file_handlers/javascript.rs
Comment (ematipico) [inline crates/biome_service/src/file_handlers/javascript.rs:665]: I don't think we need to check the applicability, unless there's some logic I am missing

File: crates/biome_service/src/file_handlers/graphql.rs
Comment (ematipico) [inline crates/biome_service/src/file_handlers/graphql.rs:533]:
```suggestion
                        // TODO: implement once a GraphQL suppression action is available
```

File: crates/biome_service/src/file_handlers/css.rs
Comment (ematipico) [inline crates/biome_service/src/file_handlers/css.rs:558]:
```suggestion
                        // TODO: to implement
```

File: crates/biome_cli/tests/snapshots/main_cases_suppressions/suppress_only_ok.snap
Comment (ematipico) [inline crates/biome_cli/tests/snapshots/main_cases_suppressions/suppress_only_ok.snap:10]: Nice!! 😍

File: crates/biome_cli/src/execute/mod.rs
Comment (ematipico) [inline crates/biome_cli/src/execute/mod.rs:309]: I think we don't need this function anymore. We need to update `as_fix_file_mode` to return the new variant `FixFileMode::ApplySuppresion`

**Comment by anthonyshew** [comment]
Thanks, @ematipico! Addressed your comments. 👍

**Comment by anthonyshew** [comment]
@ematipico, just checked back in on this PR and had to resolve some conflicts. Is there anything else you need me to do to get merged? I'd love to get started on the custom descriptions but don't want to be working against myself. 😄

**Comment by ematipico** [comment]
@anthonyshew 

You don't need to do anything now. Since this is a feature, we need to wait a bit. We're still merging bug fixes. Once we're ready, we'll merge it

**Comment by anthonyshew** [comment]
Ah, okay, no worries. Sorry, should've put that together.

Would it be wise for me to start trying to work on the custom explanations or should I hold back on that?

**Review by ematipico — APPROVED** [review]

**Pull request merged**
